### PR TITLE
Add core API and backend service design doc

### PR DESCRIPTION
**Pull request by Jeffwan**
## Why are these changes needed?

Per Ali's request in https://github.com/ray-project/kuberay/pull/93, It's better to write a one page summary about the proposed change and get it approved before any implementation.

## Related issue number
#53 #54 

/cc @akanso @chenk008 @caitengwei @chaomengyuan 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: docs/design/protobuf-grpc-service.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  uint32 memory = 3;
+  // Number of gpus
+  uint32 gpu = 4;
+  // The detail gpu accelerator type
+  string gpu_accelerator = 5;
+}
+
+// ImageTemplate can be used by worker group and workspce.
+// They can be distinguish by different entrypoints
+message ImageTemplate {
```
Comment (Jeffwan) [inline docs/design/protobuf-grpc-service.md:116]: I think ray-operator should not be aware of the imageTemplate. it just need `image` to build the cluster.  `ImageTemplate` corresponding services can be used to build images separately. 


1. User who does not have docker experience can build images using this service. 
2. User can query the image result built by the service
3. User can use that image result to build ray cluster.

**Review by chenk008 — APPROVED** [review]

**Comment by akanso** [comment]
The functionality of exposing a simpler interface to the users other than the K8s interface is very valuable to none K8s users.
GRPC is a generic interface, many clients in GoLang, C#, Python etc. can use it.

It is not clear from the design Doc if the operator code itself is extended to support this feature? 

I think an alternative design, is to run the GPRC server as an independent container (e.g. side-car container), that is capable of processing the GRPC requests and create RayCluster Custom Resources that the Ray Operator can read and process.

client --> GRPC Server --> [created Custom Resources] <-- Ray Operator (reads CR and accordingly performs CRUD )

**Comment by Jeffwan** [comment]
@akanso

> It is not clear from the design Doc if the operator code itself is extended to support this feature?
I see.  I think the gap is I should add more details on the operator interactions and deployment topology. 

> I think an alternative design, is to run the GPRC server as an independent container (e.g. side-car container), that is capable of processing the GRPC requests and create RayCluster Custom Resources that the Ray Operator can read and process.
> client --> GRPC Server --> [created Custom Resources] <-- Ray Operator (reads CR and accordingly performs CRUD )

exactly. It would be a separate pod and the path is like what you described. Let me add more details and diagrams

**Comment by Jeffwan** [comment]
@akanso I add a short summary to talk about deployment and interactions.

File: docs/design/protobuf-grpc-service.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+- The api definition should be flexible enough to support different kinds of clients (e.g. backend, cli etc).
+- This backend service underneath should leverage generate clients to interact with existing RayCluster custom resources.
+- New added components should be plugable to existing operator.
+
+## Proposal
+
+### Deployment topology and interactive flow
+
+The new gRPC service would be a individual deployment of the KubeRay control plane and user can choose to install it optinally. It will create a service and exposes endpoint to users.
```
Comment (akanso) [inline docs/design/protobuf-grpc-service.md:23]: optinally -> optionally

Comment (Jeffwan) [reply docs/design/protobuf-grpc-service.md]: nice catch and typo is fixed

**Comment by Jeffwan** [comment]
Ali reviewed the docs and overall looks good to him as well. We can merge the PR.

**Pull request merged**